CCharBuffer / UnicodeCharBuffer Skeleton & CPP Emit Structure

1stub opened the pull request:
This lays the foundation for moving towards cpp emitting and being able to use these character buffers to build up string ropes.

There isn't a whole lot going on in the cpp emission, I more so just defined the structure and build process (similar to how SMT emission is done).

CCharBuffer and UnicodeCharBuffer don't emit any necessary JavaScript, so they are not usable (yet), but they are able to be properly parsed and type checked. For the time being they cannot have more than 8 characters in their construction, but may want to make this somewhat dynamically allow up to 16 characters.

Char buffer construction looks something like so:
 -     `let cb: CCharBuffer = CCharBuffer{c'a', c'a'}; %% Ok`
 -     `let ucb: UnicodeCharBuffer = UnicodeCharBuffer{c"a", c"b", c"日", c"a"}; %% Ok`
 -     `let cb: CCharBuffer = CCharBuffer{c'aa', c'a'}; %% Fails, more than one CChar`
 -     `let ucb: UnicodeCharBuffer = UnicodeCharBuffer{c"日日", c"b"}; %% Fails, more than one UnicodeChar`
 
 The next logical step is to add the "glue" between the cpp backend and the char buffer construction interface.

BosqueLanguage commented on src/backend/bsqon/bsqonemitter.ts:
```diff
@@ -384,6 +384,12 @@ class BSQONTypeInfoEmitter {
         else if(tt instanceof MapTypeDecl) {
             alldecls.push(this.emitMapTypeDecl(tt, instantiation));
         }
+        else if(tt instanceof CCharBufferTypeDecl) {
```
These should be subtypes of PrimitiveEntity (like Int, Float, etc.)

BosqueLanguage commented on src/backend/cpp/runtime/cppruntime.hpp:
```diff
@@ -0,0 +1,45 @@
+// Useful for keeping track of path in tree iteration
+struct PathStack {
+    uint64_t bits;
+    int depth;
```
This should be the current index in the current buffer.

BosqueLanguage commented on src/backend/cpp/runtime/cppruntime.hpp:
```diff
@@ -0,0 +1,45 @@
+#pragma once
```
Lets get all of the runtime stuff in a special namespace that is an otherwise illegal Bosque namespace name -- `_CoreCpp` would probably work.

BosqueLanguage commented on src/core/xcore_cstring_exec.bsq:
```diff
@@ -3,6 +3,39 @@ namespace Core;
+    function s_ccharbuffer_create_empty() {
+        ;
+    }
+    function s_ccharbuffer_create_1(c1: CChar) {
```
This should be a builtin function:
```
function s_create_1(c1: CChar): CCharBuffer = ccharbuffer_create_1;
```

Then in the emitter we will dispatch to a "emitBuiltinFunction" routine where you can just return the string:
```
_CoreCpp::CCharBuffer::create_1(...)
```

BosqueLanguage commented on src/core/xcore_string_exec.bsq:
```diff
@@ -3,6 +3,39 @@ namespace Core;
+namespace UnicodeCharBufferOps {
+    %% Our creation methods call cpp code for handling bit ops
+    %% Once a nice interface with the cpp emitter exists these creation functions would return UnicodeCharBuffer
+    function s_unicodecharbuffer_create_empty() {        
```
Same as CString

BosqueLanguage commented on src/frontend/assembly.ts:
```diff
@@ -876,6 +876,18 @@ class MapTypeDecl extends AbstractCollectionTypeDecl {
     }
 }
 
+class CCharBufferTypeDecl extends AbstractCollectionTypeDecl {
```
Make subtype of `PrimitiveEntityTypeDecl` these are 8 char elements that may just have fewer chars. So fixed size -- not a collection.

BosqueLanguage commented on src/frontend/checker.ts:
```diff
@@ -1267,7 +1267,15 @@ class TypeChecker {
             if(arg instanceof PositionalArgumentValue) {
                 const argtype = this.checkExpression(env, arg.exp, new SimpleTypeInferContext(etype));
-                this.checkError(arg.exp.sinfo, (argtype instanceof ErrorTypeSignature) || !this.relations.isSubtypeOf(argtype, etype, this.constraints), `Argument ${i} expected type ${etype.emit()}`);
+                if(etype.tkeystr === 'CCharBuffer') { // No template typing so we just need to ensure argtype is CChar/UnicodeChar
```
This should go away as you will be calling the fixed size constructors.

BosqueLanguage commented on src/frontend/checker_relations.ts:
```diff
@@ -731,6 +731,9 @@ class TypeCheckerRelations {
         else if(decl instanceof SetTypeDecl) {
             return t.alltermargs[0];
         }
+        else if(decl instanceof CCharBufferTypeDecl || decl instanceof UnicodeCharBufferTypeDecl) { // Only possible type is CChar or UnicodeChar
```
Also remove

BosqueLanguage commented on src/frontend/closed_terms.ts:
```diff
@@ -428,6 +428,48 @@ class InstantiationPropagator {
                 }
             }
         }
+        else if(decl instanceof CCharBufferTypeDecl) {
```
Also should be removed since not a collection

BosqueLanguage commented on src/frontend/parser.ts:
```diff
@@ -5433,6 +5433,12 @@ class Parser {
             else if(name === "EventList") {
                 tdecl = new EventListTypeDecl(this.env.currentFile, sinfo, attributes, "EventList");
             }
+            else if(name === "CCharBuffer") {
```
Ditto

BosqueLanguage commented on src/frontend/parser.ts:
```diff
@@ -5481,6 +5487,9 @@ class Parser {
         else if(tdecl instanceof EventListTypeDecl) {
             this.parseOOPMembersCommonAll(false, undefined, new Set<string>("T"), undefined, undefined, tdecl.consts, tdecl.functions, undefined, tdecl.methods, undefined, undefined, undefined);
         }
+        else if(tdecl instanceof CCharBufferTypeDecl || tdecl instanceof UnicodeCharBufferTypeDecl) {
```
Ditto

BosqueLanguage commented:
Yes, deleting these (CCharBufferTypeDecl and Unicode...) is the right approach.

1stub commented:
My most recent commit, 6853191, implemented the necessary removals per feedback and added primitive constructor type checking. 

In `checker.ts`, I added a method called `checkPrimitiveConstructor()` which simply ensures the constructors for both `CCharBuffer` and `UnicodeCharBuffer` are valid.
